urllib3/poolmanager.py

```diff
@@ -40,38 +39,42 @@
 )
 
 
-def _default_key_normalizer(key_class, request_context):
+def _default_key_normalizer(request_context):
     """
-    Create a pool key of type ``key_class`` for a request.
+    Create a pool key out of a request context dictionary.
 
     According to RFC 3986, both the scheme and host are case-insensitive.
     Therefore, this function normalizes both before constructing the pool
     key for an HTTPS request. If you wish to change this behaviour, provide
     alternate callables to ``key_fn_by_scheme``.
 
-    :param key_class:
-        The class to use when constructing the key. This should be a namedtuple
-        with the ``scheme`` and ``host`` keys at a minimum.
-
     :param request_context:
         A dictionary-like object that contain the context for a request.
         It should contain a key for each field in the :class:`HTTPPoolKey`
     """
-    context = {}
-    for key in key_class._fields:
-        context[key] = request_context.get(key)
+    # Since we mutate the dictionary, make a copy first
+    context = request_context.copy()
     context['scheme'] = context['scheme'].lower()
     context['host'] = context['host'].lower()
-    return key_class(**context)
+
+    return _frozenset_from_dictionary(context)
```

Comment:
Okay, I poked at this a while last weekend, but I'm not very happy with a few problems I've encountered with the namedtuples.

The first problem I hit is that namedtuples [can't have field names that start with underscores](https://docs.python.org/dev/library/collections.html#collections.namedtuple) and most unfortunately, there are several keyword arguments that [start with underscore](https://github.com/shazow/urllib3/blob/master/urllib3/connectionpool.py#L144). This means we have to [translate these special cases](https://github.com/shazow/urllib3/pull/1016/commits/e7920be73daa459caa9d183c3d0dd8212859374d#diff-0401b3595c5b19c0b003447baaaf35d0R102) (although admittedly that code could probably be made prettier).

The other problem I hit was that there are [other connection types that expect kwargs that I didn't account for](https://github.com/shazow/urllib3/blob/master/urllib3/contrib/socks.py#L64). This could be an additional thing other PoolManagers need to handle (by altering the default pool key I guess), but I think it would break anything that's already out there.

Both these problems could be worked around, but it'll add complexity. I wanted to see what you thought before I went too far down this path.